test/metabase/api/native_query_snippet_test.clj

```diff
@@ -0,0 +1,147 @@
+                                               :creator_id  (:id rasta)
+                                               :database_id (:id db)
+                                               :description "SQL comment snippet"
+                                               :name        "comment"}]]
+    (testing "read returns all snippet fields"
+      (let [snippet-from-api ((user->client :crowberto) :get 200 (snippet-url (:id snippet)))]
+        (is (= (select-keys snippet test-snippet-fields)
+               (select-keys snippet-from-api test-snippet-fields)))))
+
+    (testing "read fails with 403 for user without read permissions"
+      (perms/revoke-permissions! (group/all-users) db)
+      (is (= "You don't have permissions to do that."
+             ((user->client :rasta) :get 403 (str "native-query-snippet/" (:id snippet))))))))
+
+;; POST /api/native-query-snippet
+(deftest create-snippet-api-test
+  (tt/with-temp* [Database [db]]
+    (testing "new snippet field validation"
+      (is (= {:errors {:content "value must be a string."}}
+             ((user->client :rasta) :post 400 (snippet-url) {})))
+
+      (is (= {:errors {:database_id "value must be an integer greater than zero."}}
+             ((user->client :rasta) :post 400 (snippet-url) {:content "NULL"})))
+
+      (is (name-schema-error? ((user->client :rasta)
+                               :post 400 (snippet-url)
+                               {:content "NULL", :database_id (:id db)})))
+
+      (is (name-schema-error? ((user->client :rasta) :post 400 (snippet-url)
+                               {:content     "NULL"
+                                :database_id (:id db)
+                                :name        " starts with a space"})))
+
+      (is (name-schema-error? ((user->client :rasta) :post 400 (snippet-url)
+                               {:content     "NULL"
+                                :database_id (:id db)
+                                :name        "contains a } character"}))))
+
+    (testing "successful create returns new snippet's data"
+      (let [snippet-input    {:name "test-snippet", :description "Just null", :content "NULL", :database_id (:id db)}
+            snippet-from-api ((user->client :crowberto) :post 200 (snippet-url) snippet-input)]
+        (is (pos? (:id snippet-from-api)))
+
+        (is (= (:database_id snippet-input)
+               (:database_id snippet-from-api)))
+
+        (is (= (:name snippet-input)
+               (:name snippet-from-api)))
+
+        (is (= (:description snippet-input)
+               (:description snippet-from-api)))
+
+        (is (= (:content snippet-input)
+               (:content snippet-from-api)))
+
+        (is (= (:id (fetch-user :crowberto))
+               (:creator_id snippet-from-api)))
+
+        (is (false? (:archived snippet-from-api)))
+
+        (is (instance? LocalDateTime (:created_at snippet-from-api)))
+
+        (is (instance? LocalDateTime (:updated_at snippet-from-api)))))
```

Comment:
if these fail it's going to be confusing to figure out what's what it would probably be better to do something like

```clj
(doseq [k [:database_id :name :description]]
  (testing k
    (is (= (get snippet-input k)
           (get snippet-from-api k)))))
```